mypy/git.py

```diff
@@ -0,0 +1,123 @@
+    """Is the given directory version-controlled with git?"""
+    return os.path.exists(os.path.join(dir, ".git"))
+
+
+def have_git() -> bool:
+    """Can we run the git executable?"""
+    try:
+        subprocess.check_output(["git", "--help"])
+        return True
+    except subprocess.CalledProcessError:
+        return False
+
+
+def get_submodules(dir: str):
+    """Return a list of all git top-level submodules in a given directory."""
+    # It would be nicer to do
+    # "git submodule foreach 'echo MODULE $name $path $sha1 $toplevel'"
+    # but that wouldn't work on Windows.
+    output = subprocess.check_output(["git", "submodule", "status"], cwd=dir)
+    # "<status><sha1> name desc"
+    # status='-': not initialized
+    # status='+': changed
+    # status='u': merge conflicts
+    for line in output.splitlines():
+        name = line.split(b" ")[1]
+        yield name.decode(sys.getfilesystemencoding())
+
+
+def git_revision(dir: str) -> bytes:
+    """Get the SHA-1 of the HEAD of a git repository."""
+    return subprocess.check_output(["git", "rev-parse", "HEAD"], cwd=dir).strip()
+
+
+def submodule_revision(dir: str, submodule: str) -> bytes:
+    """Get the SHA-1 a submodule is supposed to have."""
+    output = subprocess.check_output(["git", "ls-files", "-s", submodule], cwd=dir).strip()
+    # E.g.: "160000 e4a7edb949e0b920b16f61aeeb19fc3d328f3012 0       typeshed"
+    return output.split()[1]
+
+
+def is_dirty(dir: str) -> bool:
+    """Check whether a git repository has uncommitted changes."""
+    output = subprocess.check_output(["git", "status", "-uno", "--porcelain"], cwd=dir)
+    return output.strip() != b""
+
+
+def has_extra_files(dir: str) -> bool:
+    """Check whether a git repository has untracked files."""
+    output = subprocess.check_output(["git", "clean", "--dry-run", "-d"], cwd=dir)
+    return output.strip() != b""
+
+
+def warn_no_git_executable() -> None:
+    print("Warning: Couldn't check git integrity. "
+          "git executable not in path.", file=sys.stderr)
+
+
+def warn_dirty(dir) -> None:
+    print("Warning: git module '{}' has uncommitted changes.".format(dir),
+          file=sys.stderr)
+    print("Go to the directory", file=sys.stderr)
+    print("  {}".format(dir), file=sys.stderr)
+    print("and commit or reset your changes", file=sys.stderr)
+
+
+def warn_extra_files(dir) -> None:
+    print("Warning: git module '{}' has untracked files.".format(dir),
+          file=sys.stderr)
+    print("Go to the directory", file=sys.stderr)
+    print("  {}".format(dir), file=sys.stderr)
+    print("and add & commit your new files.", file=sys.stderr)
+
+
+def error_submodule_not_initialized(name: str, dir: str) -> None:
+    print("Submodule '{}' not initialized.".format(name), file=sys.stderr)
+    print("Please run:", file=sys.stderr)
+    print("  cd {}".format(pipes.quote(dir)), file=sys.stderr)
+    print("  git submodule init {}".format(name), file=sys.stderr)
```

Comment:
Also, it would be nicer to have all commands on a single line (as `git submodule init ... && git submodule update ...`) for easier copy-paste (but only on a POSIX system). I guess it's safe to assume a sufficiently bash-like shell?